### PR TITLE
Oracle Service

### PR DESCRIPTION
**Pull request by shargon**
- Added Oracle Service
- Added Filters
- Added comunication method between OracleResponseTx and OracleUserTx

*TODO*
*- Decide the output (transaction, or message)*

### CODE REVIEW

**Comment by shargon** [comment]
@erikzhang @neo-project/core @neo-project/ngd-shanghai  Please could you review it and go to the next step?

- Multisign address between the Oracle Nodes (define `m`)
- Transaction Types (Normal,OracleTx and OracleResponseTx)
- Transaction Attribute -> `DependsOn` (only for OracleResponseTx)
- MemoryPool changes
- Comunication between these two TX
- Filters

**Comment by Tommo-L** [comment]
Agree, let's move forward. Oracle is really not an easy thing.

File: src/neo/Oracle/OracleService.cs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+namespace Neo.Oracle
+{
+    public class OracleService : UntypedActor
+    {
+        /// <summary>
+        /// A flag for know if the pool was ordered or not
+        /// </summary>
+        private bool _itsDirty = false;
+        private long _isStarted = 0;
```
Comment (Tommo-L) [inline src/neo/Oracle/OracleService.cs:23]: _isDirty?

File: src/neo/Oracle/Protocols/Https/OracleHttpsProtocol.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+                if (!ret.IsFaulted)
+                {
+                    if (!OracleService.FilterResponse(ret.Result, request.Filter, out string filteredStr))
+                    {
+                        return OracleResponse.CreateError(request.Hash, OracleResultError.FilterError);
+                    }
+
+                    return OracleResponse.CreateResult(request.Hash, filteredStr);
+                }
```
Comment (Tommo-L) [inline src/neo/Oracle/Protocols/Https/OracleHttpsProtocol.cs:137]: If possible, could we optimize here, as there are too many if branchs to find the normal statement. It's just my personal habit of putting all the error handling together and putting the normal processing at the end. ☺️

File: src/neo/IO/Helper.cs [hunk 3]
```diff
@@ -322,6 +322,15 @@ public static void WriteFixedString(this BinaryWriter writer, string value, int
...
         }
 
+        public static void WriteArray<T>(this BinaryWriter writer, T[] value) where T : class, ISerializable
+        {
+            writer.WriteVarInt(value.Length);
+            foreach (var item in value)
+            {
+                item.Serialize(writer);
+            }
+        }
```
Comment (erikzhang) [inline src/neo/IO/Helper.cs:332]: https://github.com/neo-project/neo/blob/98ec16958b9ee4e8e84660c88bcd8a1e01be894b/src/neo/IO/Helper.cs#L271-L278

There is a method with the same function, and it should not be written again.

**Comment by shargon** [comment]
Merged in order to continue with:

- Multisign address between the Oracle Nodes (define m)
- Transaction Types (Normal,OracleTx and OracleResponseTx)
- Transaction Attribute -> DependsOn (only for OracleResponseTx)
- MemoryPool changes
- Create Asserts for wallets if the user want to

If you want to review oracle, you can compare it with master here
https://github.com/neo-project/neo/compare/master...neo-project:oracles?expand=1

**Pull request merged**